### ISSUE TITLE
cgroups version detection fails when systemd.unified_cgroup_hierarchy=1 and default-hierarchy=hybrid

### ISSUE THREAD

**Issue by ilyam8** [netdata/netdata#11094]
##### Bug report summary

ssia

[cgroups_try_detect_version](https://github.com/netdata/netdata/blob/037584392c9ddb0fecae6ded7c258da783df5aec/collectors/cgroups.plugin/sys_fs_cgroup.c#L138-L198) fails to detect correct Cgroup version and fallbacks to v1.

The problematic part

https://github.com/netdata/netdata/blob/037584392c9ddb0fecae6ded7c258da783df5aec/collectors/cgroups.plugin/sys_fs_cgroup.c#L164-L175

In my case i get `systemd_setting == SYSTEMD_CGROUP_HYBRID`. Seems like our approach is wrong.

---

<details>
<summary>ls -l /sys/fs/cgroup/ [Click]</summary>

```CMD
[ilyam@pc netdata]$ ls -l /sys/fs/cgroup/
total 0
-r--r--r--  1 root root 0 мая  5 13:40 cgroup.controllers
-rw-r--r--  1 root root 0 мая  5 15:00 cgroup.max.depth
-rw-r--r--  1 root root 0 мая  5 15:00 cgroup.max.descendants
-rw-r--r--  1 root root 0 мая  5 13:40 cgroup.procs
-r--r--r--  1 root root 0 мая  5 15:00 cgroup.stat
-rw-r--r--  1 root root 0 мая  5 13:40 cgroup.subtree_control
-rw-r--r--  1 root root 0 мая  5 15:00 cgroup.threads
-rw-r--r--  1 root root 0 мая  5 15:00 cpu.pressure
-r--r--r--  1 root root 0 мая  5 13:40 cpuset.cpus.effective
-r--r--r--  1 root root 0 мая  5 13:40 cpuset.mems.effective
-r--r--r--  1 root root 0 мая  5 15:00 cpu.stat
drwxr-xr-x  2 root root 0 мая  5 14:11 dev-hugepages.mount
drwxr-xr-x  2 root root 0 мая  5 14:11 dev-mqueue.mount
drwxr-xr-x  2 root root 0 мая  5 13:40 init.scope
-rw-r--r--  1 root root 0 мая  5 15:00 io.cost.model
-rw-r--r--  1 root root 0 мая  5 15:00 io.cost.qos
-rw-r--r--  1 root root 0 мая  5 15:00 io.pressure
-r--r--r--  1 root root 0 мая  5 15:00 io.stat
-r--r--r--  1 root root 0 мая  5 15:00 memory.numa_stat
-rw-r--r--  1 root root 0 мая  5 15:00 memory.pressure
-r--r--r--  1 root root 0 мая  5 15:00 memory.stat
drwxr-xr-x  2 root root 0 мая  5 14:11 proc-sys-fs-binfmt_misc.mount
drwxr-xr-x  2 root root 0 мая  5 14:11 sys-fs-fuse-connections.mount
drwxr-xr-x  2 root root 0 мая  5 14:11 sys-kernel-config.mount
drwxr-xr-x  2 root root 0 мая  5 14:11 sys-kernel-debug.mount
drwxr-xr-x  2 root root 0 мая  5 14:11 sys-kernel-tracing.mount
drwxr-xr-x 33 root root 0 мая  5 14:43 system.slice
drwxr-xr-x  3 root root 0 мая  5 14:11 user.slice
```

</details>

<details>
<summary>Relevant logs [Click]</summary>

```CMD
2021-05-05 14:43:49: netdata INFO  : PLUGIN[cgroups] : use unified cgroups false
2021-05-05 14:43:49: netdata ERROR : PLUGIN[cgroups] : CGROUP: cannot find cpuacct mountinfo. Assuming default: /sys/fs/cgroup/cpuacct
2021-05-05 14:43:49: netdata ERROR : PLUGIN[cgroups] : CGROUP: cannot find cpuset mountinfo. Assuming default: /sys/fs/cgroup/cpuset
2021-05-05 14:43:49: netdata ERROR : PLUGIN[cgroups] : CGROUP: cannot find blkio mountinfo. Assuming default: /sys/fs/cgroup/blkio
2021-05-05 14:43:49: netdata ERROR : PLUGIN[cgroups] : CGROUP: cannot find memory mountinfo. Assuming default: /sys/fs/cgroup/memory
2021-05-05 14:43:49: netdata ERROR : PLUGIN[cgroups] : CGROUP: cannot find devices mountinfo. Assuming default: /sys/fs/cgroup/devices
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: cannot read directory '/sys/fs/cgroup/cpuacct' (errno 2, No such file or directory)
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: disabled cpu statistics.
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: cannot read directory '/sys/fs/cgroup/blkio' (errno 2, No such file or directory)
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: disabled blkio statistics.
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: cannot read directory '/sys/fs/cgroup/memory' (errno 2, No such file or directory)
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: disabled memory statistics.
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: cannot read directory '/sys/fs/cgroup/devices' (errno 2, No such file or directory)
2021-05-05 14:44:00: netdata ERROR : MAIN : CGROUP: disabled devices statistics.
```

</details>

##### OS / Environment

```CMD
Linux pc 5.10.32-1-MANJARO #1 SMP Wed Apr 21 14:54:10 UTC 2021 x86_64 GNU/Linux
```

<details>
<summary>release [Click]</summary>

```CMD
/etc/arch-release:Manjaro Linux
/etc/lsb-release:DISTRIB_ID=ManjaroLinux
/etc/lsb-release:DISTRIB_RELEASE=21.0.3
/etc/lsb-release:DISTRIB_CODENAME=Ornara
/etc/lsb-release:DISTRIB_DESCRIPTION="Manjaro Linux"
/etc/manjaro-release:Manjaro Linux
/etc/os-release:NAME="Manjaro Linux"
/etc/os-release:ID=manjaro
/etc/os-release:ID_LIKE=arch
/etc/os-release:BUILD_ID=rolling
/etc/os-release:PRETTY_NAME="Manjaro Linux"
/etc/os-release:ANSI_COLOR="32;1;24;144;200"
/etc/os-release:HOME_URL="https://manjaro.org/"
/etc/os-release:DOCUMENTATION_URL="https://wiki.manjaro.org/"
/etc/os-release:SUPPORT_URL="https://manjaro.org/"
/etc/os-release:BUG_REPORT_URL="https://bugs.manjaro.org/"
/etc/os-release:LOGO=manjarolinux
```

</details>

##### Netdata version

```CMD
Version: netdata v1.30.1-113-g037584392
```

<details>
<summary>buildinfo [Click]</summary>

```CMD
Configure options:  '--prefix=/opt/netdata/usr' '--sysconfdir=/opt/netdata/etc' '--localstatedir=/opt/netdata/var' '--libexecdir=/opt/netdata/usr/libexec' '--libdir=/opt/netdata/usr/lib' '--with-zlib' '--with-math' '--with-user=netdata' '--disable-cloud' 'CFLAGS=-O2' 'LDFLAGS='
Features:
    dbengine:                YES
    Native HTTPS:            YES
    Netdata Cloud:           NO (by user request)
    TLS Host Verification:   YES
Libraries:
    jemalloc:                NO
    JSON-C:                  YES
    libcap:                  YES
    libcrypto:               YES
    libm:                    YES
    LWS:                     YES shared-lib
    mosquitto:               YES
    tcalloc:                 NO
    zlib:                    YES
Plugins:
    apps:                    YES
    cgroup Network Tracking: YES
    CUPS:                    YES
    EBPF:                    YES
    IPMI:                    NO
    NFACCT:                  NO
    perf:                    YES
    slabinfo:                YES
    Xen:                     NO
    Xen VBD Error Tracking:  NO
Exporters:
    AWS Kinesis:             NO
    GCP PubSub:              NO
    MongoDB:                 NO
    Prometheus Remote Write: YES
```

</details>

##### Installation method

From source.

##### Component Name

`collectors/cgroups`

##### Steps To Reproduce

- enable Cgroups V2
  - add _systemd.unified_cgroup_hierarchy=1_ to the `GRUB_CMDLINE_LINUX_DEFAULT` (`/etc/default/grub`)
  - update-grub
  - reboot

- start several containers

##### Expected behavior

Cgroups version detection works, it correctly detects Cgroups version when Cgroups V2 is enabled.

**Comment by ilyam8** [comment]
[IsCgroup2UnifiedMode](https://github.com/opencontainers/runc/blob/b23315bdd99c388f5d0dd3616188729c5a97484a/libcontainer/cgroups/utils.go#L35-L51) function from the [opencontainers/runc/libcontainer](https://github.com/opencontainers/runc/tree/master/libcontainer#libcontainer).

```GO
const (
	unifiedMountpoint = "/sys/fs/cgroup"
)

// IsCgroup2UnifiedMode returns whether we are running in cgroup v2 unified mode.
func IsCgroup2UnifiedMode() bool {
	isUnifiedOnce.Do(func() {
		var st unix.Statfs_t
		err := unix.Statfs(unifiedMountpoint, &st)
		if err != nil {
			if os.IsNotExist(err) && userns.RunningInUserNS() {
				// ignore the "not found" error if running in userns
				logrus.WithError(err).Debugf("%s missing, assuming cgroup v1", unifiedMountpoint)
				isUnified = false
				return
			}
			panic(fmt.Sprintf("cannot statfs cgroup root: %s", err))
		}
		isUnified = st.Type == unix.CGROUP2_SUPER_MAGIC
	})
	return isUnified
}
```

I think we could use the same logic:
 - statfs cgroup mountpoint
 - check the type

I tested this on several servers (some have cgroup v1, some v2) and it works perfectly.

**Comment by vlvkobal** [comment]
Here is a quick dirty fix - #11102. We should rewrite cgroups plugin, though. It should support hybrid cgroup configurations and do correct detection based on mountpoints - no loose magic.

**Comment by ilyam8** [comment]
Quick - agree, dirty - disagree 😉
